staging/src/k8s.io/client-go/rest/config.go

```diff
@@ -292,12 +287,24 @@ func InClusterConfig() (*Config, error) {
 		tlsClientConfig.CAFile = rootCAFile
 	}
 
-	return &Config{
-		// TODO: switch to using cluster DNS.
-		Host:            "https://" + net.JoinHostPort(host, port),
+	config := &Config{
 		BearerToken:     string(token),
 		TLSClientConfig: tlsClientConfig,
-	}, nil
+	}
+
+	// Get the kubernetes service port using the SRV record. If that fails, fallback to use
+	// the environment variables.
+	if _, addresses, err := net.LookupSRV("https", "tcp", "kubernetes.default.svc"); err == nil && len(addresses) > 0 {
```

Comment:
This also breaks any pod that has DNS set to a different resolution policy, or hostNetwork pods.  Which means it's potentially a security risk to change this.  We don't want to break things using hostNetwork.